lib/std/parker.zig

```diff
@@ -0,0 +1,381 @@
+    pub fn init() LinuxParker {
+        return LinuxParker{};
+    }
+    pub fn deinit(self: *LinuxParker) void {}
+
+    pub fn unpark(self: *LinuxParker, ptr: *const u32) void {
+        const rc = linux.futex_wake(@ptrCast(*const i32, ptr), linux.FUTEX_WAKE | linux.FUTEX_PRIVATE_FLAG, 1);
+        assert(linux.getErrno(rc) == 0);
+    }
+
+    pub fn park(self: *LinuxParker, ptr: *const u32, expected: u32) void {
+        const value = @intCast(i32, expected);
+        while (@atomicLoad(u32, ptr, .Acquire) == expected) {
+            const rc = linux.futex_wait(@ptrCast(*const i32, ptr), linux.FUTEX_WAIT | linux.FUTEX_PRIVATE_FLAG, value, null);
+            switch (linux.getErrno(rc)) {
+                0, linux.EAGAIN => return,
+                linux.EINTR => continue,
+                linux.EINVAL => unreachable,
+                else => unreachable,
+            }
+        }
+    }
+};
+
+const WindowsParker = struct {
+    waiters: u32,
+
+    pub fn init() WindowsParker {
+        return WindowsParker{ .waiters = 0 };
+    }
+    pub fn deinit(self: *WindowsParker) void {}
+
+    pub fn unpark(self: *WindowsParker, ptr: *const u32) void {
+        switch (Backend.get().*) {
+            .WaitAddress => |*backend| backend.unpark(ptr, &self.waiters),
+            .KeyedEvent => |*backend| backend.unpark(ptr, &self.waiters),
+        }
+    }
+
+    pub fn park(self: *WindowsParker, ptr: *const u32, expected: u32) void {
+        switch (Backend.get().*) {
+            .WaitAddress => |*backend| backend.park(ptr, expected, &self.waiters),
+            .KeyedEvent => |*backend| backend.park(ptr, expected, &self.waiters),
+        }
+    }
+
+    const Backend = union(enum) {
+        WaitAddress: WaitAddress,
+        KeyedEvent: KeyedEvent,
+
+        var backend = std.lazyInit(Backend);
+
+        fn get() *const Backend {
+            return backend.get() orelse {
+                // Statically linking to the KeyedEvent functions should mean its supported.
+                // TODO: Maybe add a CreateSemaphore backend for systems older than Windows XP ?
+                backend.data = WaitAddress.init() 
+                    orelse KeyedEvent.init()
+                    orelse unreachable; 
+                backend.resolve();
+                return &backend.data;
+            };
+        }
+
+        const WaitAddress = struct {
+            WakeByAddressSingle: stdcallcc fn(Address: *const c_void) void,
+            WaitOnAddress: stdcallcc fn (
+                Address: *const c_void,
+                CompareAddress: *const c_void,
+                AddressSize: windows.SIZE_T,
+                dwMilliseconds: windows.DWORD,
+            ) windows.BOOL,
+
+            fn init() ?Backend {
+                const dll_name = c"api-ms-win-core-synch-l1-2-0";
+                const dll = windows.kernel32.GetModuleHandleA(dll_name)
+                    orelse windows.kernel32.LoadLibraryA(dll_name)
+                    orelse return null;
```

Comment:
There are 2 things that make me prefer the KeyedEvent backend here:

1. It does not require this runtime dynamic library loading step.
2. NtDll is the most basic dependency; in general we're trying to move to preferring NtDll to other higher level APIs when possible. Everything already depends on NtDll, so this dependency is "free". However this would be the first dependency on `api-ms-win-core-synch-l1-2-0`. It's also likely that api-ms-win-core-synch-l1-2-0 is implemented by calling NtDll functions anyway.

Would it be reasonable to toss the `WaitOnAddress` backend and rely exclusively on `WaitOnKeyedEvent`?

Reply:
Sure. The use was about of trying to use newer APIs, but according to [these](https://devblogs.microsoft.com/oldnewthing/20160826-00/?p=94185) [articles](http://joeduffyblog.com/2006/11/28/windows-keyed-events-critical-sections-and-new-vista-synchronization-features/), it seems as though WaitOnAddress *was* a buffered layer over Keyed Events like you said

Reply:
There's a neat tool you can use, [ProcMon](https://docs.microsoft.com/en-us/sysinternals/downloads/procmon) that will reveal what functions are actually being called. Blew my mind when I discovered it.

Reply:
Oh, another question on the lazy initialization of the Keyed Event: should it still panic on failure to create a "event handle" or is there a better way to handle that?

Reply:
This is a difficult consideration. To gain some insight on this matter it would be nice to look at the source code of Windows. That's not possible, but it is possible to look at the source code of Wine and ReactOS. Here's the function from Wine:

```c++
NTSTATUS WINAPI NtCreateKeyedEvent( HANDLE *handle, ACCESS_MASK access,
                                    const OBJECT_ATTRIBUTES *attr, ULONG flags )
{
    NTSTATUS ret;
    data_size_t len;
    struct object_attributes *objattr;

    if ((ret = alloc_object_attributes( attr, &objattr, &len ))) return ret;

    SERVER_START_REQ( create_keyed_event )
    {
        req->access = access;
        wine_server_add_data( req, objattr, len );
        ret = wine_server_call( req );
        *handle = wine_server_ptr_handle( reply->handle );
    }
    SERVER_END_REQ;

    RtlFreeHeap( GetProcessHeap(), 0, objattr );
    return ret;
}
```

The real Windows code is probably a lot different. But if this function can fail in Wine, it's pretty likely that it can also fail in Windows. And here you can see that it does a heap memory allocation, and bails out early if the allocation can fail.

To me this means that not even `@panic` is really an acceptable way to handle an error from this API. Typically the way that I would want to set this up is to have `init()` return a possible error. This makes the hot path more deterministic. However in the case of "thread parking", park() is already invoking non-determinism. So I think it is OK to do some lazy initialization, at least as far as determinism is concerned.

But that still leaves us with handling the rare possibility of failure.

At a high abstraction level, mutex operations must always succeed. I think that is a reasonable API to expect and to provide. So we will have to handle failure at a low level of using this keyed event thing, and still satisfy the semantics of the API.

One trick up our sleeves could be if park() is defined to have spurious wakeups. In this case, we could handle the rare one-time initialization failure of `NtCreateKeyedEvent` by sleeping for some small duration of time and then returning, or even simply returning without interacting with the operating system at all.

This would be reasonable because it would be extremely rare for this to happen, and if it did, correctness would be achieved at the cost of performance, in a codepath which is understood to be non-deterministic.

Reply:
For some extra context, the "park" naming scheme was after Webkit's `WTF::Lock`. Also that's a nice workaround to ensure correctness. Sleeping incrementally longer up to a threads time slice like `SpinLock.Backoff` may be close to ideal for continuous failure. Does the solution already work without making park() have spurious wake-ups since it rechecks the condition in a while loop after blocking?

Reply:
> Does the solution already work without making park() have spurious wake-ups since it rechecks the condition in a while loop after blocking?

Yes I believe the usage of park() already handles spurious wakeups, so that's what makes this solution desirable- it's more just clarifying the API and avoiding calling `@panic()`.

I'm a bit confused at the phrasing of "making park() have spurious wake-ups" because I'm not suggesting to write any more code; just to document in the API of the function that it is allowed to spuriously wake up. (and then remove the part where it calls `@panic()`).

As for something more involved like incrementally increasing sleeps - I won't tell you not to do it, but I suspect that it won't be worth doing something as complex as this since this will be a very rare situation.